src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16TextEncodingLE.cs

```diff
@@ -7,113 +7,145 @@ namespace System.Text.Utf16
 {
     class Utf16TextEncodingLE : TextEncoder
     {
-        public override bool TryEncodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> buffer, out int bytesWritten)
+        #region Encoding Constants
+
+        // TODO: Some of these members are needed only in Utf16LittleEndianEncoder.
+        //       Should we add the usage of them to UnicodeCodePoint class and merge this class with it?
+        private const uint Utf16HighSurrogateFirstCodePoint = 0xD800;
+        private const uint Utf16HighSurrogateLastCodePoint = 0xDFFF;
+        private const uint Utf16LowSurrogateFirstCodePoint = 0xDC00;
+        private const uint Utf16LowSurrogateLastCodePoint = 0xDFFF;
+
+        private const uint Utf16SurrogateRangeStart = Utf16HighSurrogateFirstCodePoint;
+        private const uint Utf16SurrogateRangeEnd = Utf16LowSurrogateLastCodePoint;
+
+        // To get this to compile with dotnet cli, we need to temporarily un-binary the magic values
+        private const byte b0000_0111U = 7;
+        private const byte b0000_1111U = 15;
+        private const byte b0001_1111U = 31;
+        private const byte b0011_1111U = 63;
+        private const byte b0111_1111U = 127;
+        private const byte b1000_0000U = 128;
+        private const byte b1100_0000U = 192;
+        private const byte b1110_0000U = 224;
+        private const byte b1111_0000U = 240;
+        private const byte b1111_1000U = 248;
+
+        #endregion Encoding Constants
+
+        #region Decoding implementation
+
+        public override bool TryDecode(ReadOnlySpan<byte> data, out string text, out int bytesConsumed)
         {
-            bytesWritten = 0;
-            int justWritten;
-            foreach (var cp in new Utf8String(utf8).CodePoints)
-            {
-                if (!Utf16LittleEndianEncoder.TryEncodeCodePoint(cp, buffer.Slice(bytesWritten), out justWritten))
-                {
-                    bytesWritten = 0;
-                    return false;
-                }
-                bytesWritten += justWritten;
-            }
+            var utf16 = data.Cast<byte, char>();
```

Comment:
This cast is not portable. In fact this method will be removed. You need to use AsBytes when casting to Span<byte> and NonPortableCast when casting to anything else.